### PR TITLE
Add additional Raft performance metrics

### LINKED ISSUE #11418: Add additional performance metrics
**Description**

Please add more I/O performance metrics, namely:

- [x] Read IOPS over time
- [x] Write IOPS over time
- [x] Size distribution in KB of the Raft append requests
- [x] Time segment allocation isolated from segment creation
- [x] Time distribution of the meta store flush, notably setLastWrittenIndex, separated from the normal flush
- [x] Append data rate, i.e. rate in KiB of how fast we append data on followers
- [x] Append rate, i.e. rate of how many entries we append on followers per second
- [x] Commit rate, i.e. rate of how many entries we commit per second
- [x] Journal append latency
- [x] Journal append data rate
- [x] Number of records not replicated
- [x] Number of records not committed
- [ ] Add p95 and p99 quantile to most latency measuring diagrams

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR adds some new performance metrics for Raft, and some new graphs to the dashboard.

In the I/O section:

- Read and write IOPS graphs

In the Raft section:

- Latency distribution of segment pre-allocation (isolated from segment creation)
- Latency distribution of setting the last written index in the meta store
- The rate at which data is appended to followers in KiB per partition per follower
- The rate at which entries are appended to followers per partition per follower
- The rate at which entries are committed per partition
- The rate at which entries are appended to the journal
- The rate at which data is appended to the journal
- The latency distribution of appends to the journal
- The count of non committed records on the leader
- The count of non replicated records per follower

You can see a test running under `np-test` with [this dashboard](https://grafana.dev.zeebe.io/goto/riHXMOAVk?orgId=1).

I did not add the data rate at which we commit entries - this may be useful, but it wasn't a low hanging fruit to find that out. Since I mostly want to compare how fast we commit vs how fast we append, I figured comparing the rate by counting entries achieves the same.

## Related issues

closes #11418 

## Definition of Ready

Please check the items that apply, before requesting a review.

You can find more details about these items in our wiki page about [Pull Requests and Code Reviews](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

* [ ] I've reviewed my own code
* [ ] I've written a clear changelist description
* [ ] I've narrowly scoped my changes
* [ ] I've separated structural from behavioural changes

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by npepinpe** [comment]
Note that about 1700 lines come from the dashboard itself, so this PR is not as big as it seems.

**Comment by npepinpe** [comment]
BTW the dashboard contains the changes in #11512, so I'll wait for that to be merged first.

**Comment by npepinpe** [comment]
There was a flaky test initially, and it seems the check isn't updated even if the job is re-run.

**Comment by npepinpe** [comment]
I've rebased. The new benchmark is running under np-test now, and it includes data for the sequencer metrics FYI.

File: atomix/cluster/src/main/java/io/atomix/raft/metrics/LeaderMetrics.java
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/metrics/LeaderMetrics.java:38]: :wrench: The value is not really the rate. It is the total count. When we visualize it, we use the `rate()` function. So it might be more accurate to specify it as the total count in the help description.

File: atomix/cluster/src/main/java/io/atomix/raft/metrics/LeaderMetrics.java
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/metrics/LeaderMetrics.java:46]: :wrench: same here. It is count.

File: atomix/cluster/src/main/java/io/atomix/raft/metrics/LeaderMetrics.java
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/metrics/LeaderMetrics.java:61]: :wrench: Same here.

File: atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderAppender.java
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderAppender.java:571]: :x: This must be also updated on append. Otherwise, if nothing is committed we won't see it reflected in the metrics.

File: atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java:236]: :wrench: Not sure if this is really necessary here. On becoming leader, follower's state is reset so the matchIndex will be 0.  So the gap repeated by the metrics will be always high immediately after leader change, even though the followers are almost up-to-date.

File: atomix/cluster/src/main/java/io/atomix/raft/metrics/RaftRoleMetrics.java
Comment (deepthidevaki) [inline atomix/cluster/src/main/java/io/atomix/raft/metrics/RaftRoleMetrics.java:78]: :question: This is caching for optimization as per prometheus recommendation, right?

**Review by deepthidevaki — APPROVED** [review]
:rocket: Thanks. Looks good to me. Only some minor comments.

File: atomix/cluster/src/main/java/io/atomix/raft/metrics/RaftRoleMetrics.java
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/metrics/RaftRoleMetrics.java:78]: Exactly, see also https://github.com/camunda/zeebe/issues/11267

File: atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java
Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java:236]: I added it for the reset on cluster restart or something like that. But you're right that it will always `lastIndex`, so it's not that much more useful - I guess what I wanted is to reset it to 0. Should we just do that?

Comment (deepthidevaki) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java:236]: Setting to 0 would be fine as well. Or we don't have to set anything at all. Then grafana shows no value until the first AppendRequest is acknowledged, and that is ok I guess.

Comment (npepinpe) [reply atomix/cluster/src/main/java/io/atomix/raft/roles/LeaderRole.java:236]: Let's remove it

**Comment by npepinpe** [comment]
There will be a merge conflict with https://github.com/camunda/zeebe/pull/11521, so I'll wait until that's merged before merging this.

**Comment by npepinpe** [comment]
bors merge

**Pull request merged**
